### PR TITLE
Containers: Fix image URL logic

### PR DESCRIPTION
**Pull request by pdostal**
* Rename and fix `containers::urls::get_container_url_from_var()` (previously `get_urls_from_var()`)
 * Rename `CONTAINER_IMAGES_TO_TEST` to `CONTAINER_IMAGE_TO_TEST`
 * Introduce `containers::urls::get_container_url_to_test()`
   * Usable in test modules where only single image is expected.
   * Not usable in `tests/containers/image.pm` where multiple images of various versions are expected.

- Related ticket: [poo#107776](https://progress.opensuse.org/issues/107776)
- Follow-up of: #14400
- Parallel attempt: #14434 
- Verification run: [EKS](http://pdostal-server.suse.cz/tests/13764), [EKS with specified image](http://pdostal-server.suse.cz/tests/13765), [Docker with specified image](http://pdostal-server.suse.cz/tests/13766), [Docker](http://pdostal-server.suse.cz/tests/13767), [Docker with multiple SLE image versions](http://pdostal-server.suse.cz/tests/13753)

### CODE REVIEW

File: lib/containers/urls.pm [hunk 1]
```diff
@@ -293,6 +294,16 @@ sub get_suse_container_urls {
     return (\@untested_images, \@released_images);
 }
 
+# Get single image from CONTAINER_IMAGE_TO_TEST or untested image from get_suse_container_urls()
+sub get_container_url_to_test {
```
Comment (grisu48) [inline lib/containers/urls.pm:298]:
```suggestion
sub get_container_url_to_test {
sub get_container_image_to_test {
```

Maybe we could align the naming of the function with the naming of the variable? WDYT?

File: lib/containers/urls.pm [hunk 2]
```diff
@@ -262,12 +263,12 @@ sub supports_image_arch {
     (grep { $_ eq $arch } @{$images_uri{$distri}{$version}{available_arch}}) ? 1 : 0;
 }
 
-# Returns a tuple of image urls.
+# Returns a tuple of single image url.
```
Comment (grisu48) [inline lib/containers/urls.pm:266]:
```suggestion
# Returns an array with the contents of the given variable or undef, if the var is not set
```
Sorry, I'm nitpicking here :slightly_smiling_face: 

AFAIK perl has no tuples, also it does not return empty but undef if the var is not set. I made a suggestion how this can be addressed.

**Review by grisu48 — COMMENTED** [review]
Good work Pavel! I have two suggestions for improvement but overall this makes more sense now.

**Review by ilausuch — APPROVED** [review]
It's a ready good job.

File: lib/containers/urls.pm (hunk 2 above)
Comment (pdostal) [reply lib/containers/urls.pm:266]: Nice, thank you :smile:

Comment (b10n1k) [reply lib/containers/urls.pm]: more specifically a `ref to array`. one question. @pdostal  Why dont you use perlpod format?

File: lib/containers/urls.pm [hunk 3]
```diff
@@ -262,12 +263,11 @@ sub supports_image_arch {
...
-# Returns a tuple of image urls.
-# If empty, no images available.
-sub get_urls_from_var {
+# Returns an array with the contents of the given variable or undef, if the var is not set
+sub get_container_url_from_var {
     my $var = shift;
-    my @urls = split(/,/, trim(get_var($var)));
-    return (\@urls);
+    my @urls = (trim(get_var($var)));
+    return (length(@urls[0]) > 0) ? (\@urls) : undef;
```
Comment (b10n1k) [inline lib/containers/urls.pm:270]: `s/length(@urls[0])/$urls[0]/` should do the same, and it is more perlist way

File: lib/containers/urls.pm [hunk 4]
```diff
@@ -293,6 +293,16 @@ sub get_suse_container_urls {
     return (\@untested_images, \@released_images);
 }
 
+# Get single image from CONTAINER_IMAGE_TO_TEST or untested image from get_suse_container_urls()
+sub get_container_image_to_test {
+    my $images_to_test;
+    unless ($images_to_test = get_container_url_from_var('CONTAINER_IMAGE_TO_TEST')) {
```
Comment (b10n1k) [inline lib/containers/urls.pm:299]:
```suggestion
    unless (my $images_to_test = get_container_url_from_var('CONTAINER_IMAGE_TO_TEST')) {
```

File: tests/containers/image.pm [hunk 5]
```diff
@@ -28,8 +28,8 @@ sub run {
     my $versions = get_var('CONTAINER_IMAGE_VERSIONS', get_required_var('VERSION'));
     for my $version (split(/,/, $versions)) {
         my $images_to_test;
-        # Get list of images from CONTAINER_IMAGES_TO_TEST or use the default
-        unless ($images_to_test = get_urls_from_var('CONTAINER_IMAGES_TO_TEST')) {
+        # Get single image from CONTAINER_IMAGES_TO_TEST or array from get_suse_container_urls()
```
Comment (b10n1k) [inline tests/containers/image.pm:31]: i know that this is likely after @grisu48 request but i think it is good idea to point out that this is a ref to array.

**Review by b10n1k — CHANGES_REQUESTED** [review]
Some small improvements if you dont mind

File: lib/containers/urls.pm [hunk 6]
```diff
@@ -293,6 +293,16 @@ sub get_suse_container_urls {
     return (\@untested_images, \@released_images);
 }
 
+# Get single image from CONTAINER_IMAGE_TO_TEST or untested image from get_suse_container_urls()
+sub get_container_image_to_test {
+    my $images_to_test;
```
Comment (jlausuch) [inline lib/containers/urls.pm:298]: Could we do the following?
1) Remove `get_container_url_from_var` method
2) Write this method like this

```
my $image = get_var('CONTAINER_IMAGE_TO_TEST');
return $image if ($image);
($image, undef) = get_suse_container_urls();
return $image->[0]; # or similar, you can use different var if you want
```

3) For image.pm see below

File: tests/containers/image.pm [hunk 7]
```diff
@@ -28,8 +28,8 @@ sub run {
...
+        unless ($images_to_test = get_container_url_from_var('CONTAINER_IMAGE_TO_TEST')) {
```
Comment (jlausuch) [inline tests/containers/image.pm:32]: Following the previous comment, this would turn into:
```
unless (get_container_image_to_test()) {
```

File: lib/containers/urls.pm (hunk 6 above)
Comment (pdostal) [reply lib/containers/urls.pm:298]: Yes, can you please create a ticket out of this and close?

File: tests/containers/image.pm (hunk 7 above)
Comment (pdostal) [reply tests/containers/image.pm:32]: No. This does not make sense. Let me please handle this in separate pull request and please close.

File: lib/containers/urls.pm (hunk 3 above)
Comment (pdostal) [reply lib/containers/urls.pm]: Done.

File: lib/containers/urls.pm (hunk 4 above)
Comment (pdostal) [reply lib/containers/urls.pm:300]: No, this is not possible as we work with `$images_to_test` either way. I tested your suggestion tho [here](http://pdostal-server.suse.cz/tests/13776) tho.

File: tests/containers/image.pm (hunk 5 above)
Comment (pdostal) [reply tests/containers/image.pm]: Done.

File: lib/containers/urls.pm (hunk 4 above)
Comment (b10n1k) [reply lib/containers/urls.pm:300]: shit man. i missed that reused later. ignore

File: tests/containers/image.pm (hunk 7 above)
Comment (pdostal) [reply tests/containers/image.pm:32]: Or let me rephrase it: It think that this condition can be avoided completely.

File: lib/containers/urls.pm (hunk 6 above)
Comment (pdostal) [reply lib/containers/urls.pm:299]: I think [this ticket](https://progress.opensuse.org/issues/108058) is suitable.

**Review by jlausuch — APPROVED** [review]
Ok to fix the current issue that we are not testing any image.
We have https://progress.opensuse.org/issues/108058 to continue with this topic and have more time to review/suggest.

**Comment by pdostal** [comment]
All the verification runs have been restarted and are still working properly.

**Review by b10n1k — APPROVED** [review]
